Improved CountDown timer

dmgerman opened the pull request:
- Improved configuration of the notification area (size and transparency)
- Added optional icons in the menubar to start, pause and end the time, and notify of remaining time
- It is now able to deal with suspending the laptop during the timer i.e. time runs in "user" time, not "computer running" time
- added hotkeys
  - able to call timer again with last period
  - or set new period

dmgerman commented:
Is there anything else I can do to have this pull request merged? thank you.

asmagill commented:
I apologize, I told @cmsj I'd take over for him re spoons, but had some unexpected interruptions occur.  I'm hoping I can get back to this in the next day or so and run through a dry run to make sure I don't screw anything up and then start testing/merging this weekend.

dmgerman commented:
hi,

I appreciate your response.

there is no rush at all. I am happy to know somebody is thinking about it.

I have other improvements that I can share, once things get rolling.

--daniel


On Wed, Aug 30, 2023 at 1:25 PM A-Ron ***@***.***> wrote:

> I apologize, I told @cmsj <https://github.com/cmsj> I'd take over for him
> re spoons, but had some unexpected interruptions occur. I'm hoping I can
> get back to this in the next day or so and run through a dry run to make
> sure I don't screw anything up and then start testing/merging this weekend.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Hammerspoon/Spoons/pull/300#issuecomment-1699789159>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABIZ4BRGVJYNSYJPMUCQ2DXX6ORZANCNFSM6AAAAAAY3HWQ7U>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
--dmg

---
D M German
http://turingmachine.org

muescha commented on Source/CountDown.spoon/init.lua:
```diff
@@ -1,39 +1,129 @@
+end
+
+function obj:query_time_and_start()
+   local button, time = hs.dialog.textPrompt("Enter time", "in minutes", string.format("%s", obj.defaultLenMin), "Ok", "Cancel")
```
```suggestion
   local button, time = hs.dialog.textPrompt("Enter time", "in minutes", string.format("%s", obj.defaultLenMin), "OK", "Cancel")
```

muescha commented on Source/CountDown.spoon/init.lua:
```diff
@@ -1,39 +1,129 @@
+
+function obj:query_time_and_start()
+   local button, time = hs.dialog.textPrompt("Enter time", "in minutes", string.format("%s", obj.defaultLenMin), "Ok", "Cancel")
+   if button == "Ok" then
```
```suggestion
   if button == "OK" then
```

muescha commented on Source/CountDown.spoon/init.lua:
```diff
@@ -1,39 +1,129 @@
+
+-- should the icon always show?
+obj.alwaysShow=true
+-- default timer in minutes
+
+obj.defaultLenMin = 25
```
```suggestion
-- default timer in minutes
obj.defaultLenMin = 25
```

muescha commented on Source/CountDown.spoon/init.lua:
```diff
@@ -43,45 +133,145 @@ end
+function obj:create_bar_timer(minutes)
+   local mainScreen = hs.screen.primaryScreen()
+   local mainRes = mainScreen:fullFrame()
+   obj.canvas:frame({x=mainRes.x, y=mainRes.h-obj.canvasHeight, w=mainRes.w, obj.canvasHeight})
```
is here `h=` missing?

```suggestion
   obj.canvas:frame({x=mainRes.x, y=mainRes.h-obj.canvasHeight, w=mainRes.w, h=obj.canvasHeight})
```

muescha commented on Source/CountDown.spoon/init.lua:
```diff
@@ -43,45 +133,145 @@ end
-                obj:setProgress(obj.loopCount, minutes)
-            end)
-        end
+   if obj.timer then
+      -- reset current timer and do nothing
+      obj:reset()
+   else
```
what if the `minutes` value is changed?

I think it is better to cleanup if there a timer exists - or check if minutes are change
```suggestion
   if obj.timer then
      -- reset current timer and do nothing
      obj:reset()
   end
```

muescha commented on Source/CountDown.spoon/init.lua:
```diff
@@ -90,42 +280,92 @@ end
+
+      else
+         -- offset the starting time by as many seconds as we were paused
+         local pausedSeconds = obj:time_absolute_seconds() - obj.pausedAt
+         obj.startTime = obj.startTime + pausedSeconds
```
I would think this calculation is not valid if there are multiple pauses and resumes.

muescha commented on Source/CountDown.spoon/init.lua:
```diff
@@ -43,45 +133,145 @@ end
+end
+
+function obj:create_bar_timer(minutes)
+   local mainScreen = hs.screen.primaryScreen()
+   local mainRes = mainScreen:fullFrame()
```
wrong variable naming

```suggestion
   local primaryScreen = hs.screen.primaryScreen()
   local mainRes = primaryScreen:fullFrame()
```

muescha commented on Source/CountDown.spoon/init.lua:
```diff
@@ -90,42 +280,92 @@ end
     if obj.canvas:frame().h == 0 then
         -- Make the canvas actully visible
-        local mainScreen = hs.screen.mainScreen()
+        local mainScreen = hs.screen.primaryScreen()
         local mainRes = mainScreen:fullFrame()
```
wrong variable naming
```suggestion
        local primaryScreen = hs.screen.primaryScreen()
        local mainRes = primaryScreen:fullFrame()
```

muescha commented on Source/CountDown.spoon/init.lua:
```diff
@@ -1,39 +1,129 @@
+      else
+         message = string.format("Illegal number [%s]", time)
+      end
+      hs.alert.show(message, nil, hs.screen.mainScreen())
```
The canvas is changed to `primaryScreen` — should the `alert` be as well?

muescha commented on Source/CountDown.spoon/init.lua:
```diff
@@ -90,42 +280,92 @@ end
+         withdrawAfter = 100
+   }):soundName(obj.sound):send()
+
+   hs.alert.show(message, nil, hs.screen.mainScreen(), obj.alertLenMin)
```
The canvas is changed to `primaryScreen` — should the `alert` be as well?

muescha reviewed (commented):
Added some minor nitpicks.

dmgerman commented:
Thanks for the comment. There are some new features I have implemented. I'll address your comments and submit an update.

muescha commented on Source/CountDown.spoon/init.lua:
```diff
@@ -1,85 +1,638 @@
+---            the minutes that were requested.
+---            The callback is not called if timer is cancelled
+function obj:startFor(minutes, callback)
+   if not minutes  then
```
you could test minutes for type=function then minutes get the default minutes and place minutes into callback. so the minutes are optional

dmgerman replied:
do you mean: if minutes is a function, call it and use return value as length of the timer?

currently, minutes is optional, if provided, if not provided it will use defaultLenMinutes

muescha replied:
I mean that if minutes would be an optional parameter, but someone want a callback 
`function obj:startFor([minutes,] callback)`

I have seen this on some places at hammerspoon:

```lua
if type(minutes)=='function' then
	callback = minutes
	minutes = obj.defaultLenMinutes
end
```
example:
https://github.com/Hammerspoon/hammerspoon/blob/ec9a8ead1fc9074b8647c5ee4c09ba96f66a719a/extensions/host/host.lua#L89-L95

muescha commented on Source/CountDown.spoon/init.lua:
```diff
@@ -1,85 +1,638 @@
+   if not minutes  then
+      minutes = obj.defaultLenMinutes
+   end
+   if obj.timerRunning then
```
maybe do here a timer:cancel() and start over with a new timer?

dmgerman replied:
I think it is better to issue an error. The reasons is that one might accidentally try to set another timer without noticing the current one is running. I guess a better option would be to make this optional (via a variable). What do you think?

dmgerman replied:
or alternatively, ask the user.

muescha replied:
I mean that if minutes would be an optional parameter, but someone want a callback 
`function obj:startFor([minutes,] callback)`

I have seen this on some places at hammerspoon:

```lua
if type(minutes)=='function' then
	callback = minutes
	minutes = obj.defaultLenMinutes
end
```
example:
https://github.com/Hammerspoon/hammerspoon/blob/ec9a8ead1fc9074b8647c5ee4c09ba96f66a719a/extensions/host/host.lua#L89-L95

muescha commented on Source/CountDown.spoon/readme.org:
```diff
@@ -0,0 +1,162 @@
+-- configure some bindings
+   countDown:bindHotkeys({
+      startFor = {{"cmd", "ctrl", "alt"}, "P"},
+      startInteractive = {{"cmd", "ctrl", "alt", "shift"}, "P"},
```
is here this is missing:
`pauseOrResume = {{"cmd", "ctrl", "alt"}, "P"},`

dmgerman commented:
@cmsj @muescha 

Hi everybody, I have uploaded my current improvements. I have added several features since I first uploaded these changes. This is the complete summary:


     Improved CountDown timer
    
    Lots of improvements over original timer. They can be divided into several
    categories:
    
    - Improve timer handler. The original timer counted seconds via a callback.
      If the computer was suspended, the timer would have been suspended too.
      The new timer continues to run even when the computer is suspended.
    
    - Improved configurability of countdown timer
      Many of its properties can now be configured via object attributes
    
    - Added menu bar:
      A menu bar item allows to start/pause/resume/cancel a timer
    
    - Optional progress messages:
      It can optionally display messages to the screen as the timer is advancing
    
    - Improved time-up messages.
      I found that the end of the timer notifications were too subtle to be noticed.
      It now allows several ways to configure the notifications
    
    - A callback.
      User can specify a callback to the evaluated as the timer is
      started/paused/resumed/cancelled.
    
    - In addition to minutes, a timer can now be set using a time of day,

muescha reviewed (commented):
some notes

muescha commented:
your alerts are not on the same place always - is this by design?

```lua
/* 227 */ hs.alert.show(msg, obj.messageAttributes, nil, duration) -- nil
/* 291 */ hs.alert.show(message, obj.alertAttributes, screen, obj.alertLen) -- .mainScreen() / currentWin:screen()
/* 600 */ hs.alert.show(message, nil, hs.screen.mainScreen()) -- .mainScreen()
/* 630 */ hs.alert.show(message, nil, hs.screen.mainScreen()) -- .mainScreen()
```

dmgerman commented:
> your alerts are not on the same place always - is this by design?
> 
> ```lua
> /* 227 */ hs.alert.show(msg, obj.messageAttributes, nil, duration) -- nil
> /* 291 */ hs.alert.show(message, obj.alertAttributes, screen, obj.alertLen) -- .mainScreen() / currentWin:screen()
> /* 600 */ hs.alert.show(message, nil, hs.screen.mainScreen()) -- .mainScreen()
> /* 630 */ hs.alert.show(message, nil, hs.screen.mainScreen()) -- .mainScreen()
> ```

Yes. We have three types of messsage: warnings (optional), messages while setting/modifying the timer, and final alert.

cmsj commented:
Thanks @dmgerman. I'm going to merge this now, if you want to follow up on any of @muescha 's comments, feel free to open a new PR.
